pandas/tests/indexes/multi/test_integrity.py

```diff
@@ -0,0 +1,403 @@
+        1, index=pd.MultiIndex.from_product([[1, 2], range(499999)]),
+        columns=['dest'])
+    with pytest.raises(KeyError):
+        df_below_1000000.loc[(-1, 0), 'dest']
+    with pytest.raises(KeyError):
+        df_below_1000000.loc[(3, 0), 'dest']
+    df_above_1000000 = pd.DataFrame(
+        1, index=pd.MultiIndex.from_product([[1, 2], range(500001)]),
+        columns=['dest'])
+    with pytest.raises(KeyError):
+        df_above_1000000.loc[(-1, 0), 'dest']
+    with pytest.raises(KeyError):
+        df_above_1000000.loc[(3, 0), 'dest']
+
+
+def test_nan_stays_float():
+
+    # GH 7031
+    idx0 = pd.MultiIndex(levels=[["A", "B"], []],
+                         labels=[[1, 0], [-1, -1]],
+                         names=[0, 1])
+    idx1 = pd.MultiIndex(levels=[["C"], ["D"]],
+                         labels=[[0], [0]],
+                         names=[0, 1])
+    idxm = idx0.join(idx1, how='outer')
+    assert pd.isna(idx0.get_level_values(1)).all()
+    # the following failed in 0.14.1
+    assert pd.isna(idxm.get_level_values(1)[:-1]).all()
+
+    df0 = pd.DataFrame([[1, 2]], index=idx0)
+    df1 = pd.DataFrame([[3, 4]], index=idx1)
+    dfm = df0 - df1
+    assert pd.isna(df0.index.get_level_values(1)).all()
+    # the following failed in 0.14.1
+    assert pd.isna(dfm.index.get_level_values(1)[:-1]).all()
+
+
+def test_million_record_attribute_error():
+    # GH 18165
+    r = list(range(1000000))
+    df = pd.DataFrame({'a': r, 'b': r},
+                      index=pd.MultiIndex.from_tuples([(x, x) for x in r]))
+
+    with tm.assert_raises_regex(AttributeError,
+                                "'Series' object has no attribute 'foo'"):
+        df['a'].foo()
+
+
+def test_can_hold_identifiers(idx):
+    key = idx[0]
+    assert idx._can_hold_identifiers_and_holds_name(key) is True
+
+
+def test_metadata_immutable(idx):
+    levels, labels = idx.levels, idx.labels
+    # shouldn't be able to set at either the top level or base level
+    mutable_regex = re.compile('does not support mutable operations')
+    with tm.assert_raises_regex(TypeError, mutable_regex):
+        levels[0] = levels[0]
+    with tm.assert_raises_regex(TypeError, mutable_regex):
+        levels[0][0] = levels[0][0]
+    # ditto for labels
+    with tm.assert_raises_regex(TypeError, mutable_regex):
+        labels[0] = labels[0]
+    with tm.assert_raises_regex(TypeError, mutable_regex):
+        labels[0][0] = labels[0][0]
+    # and for names
+    names = idx.names
+    with tm.assert_raises_regex(TypeError, mutable_regex):
+        names[0] = names[0]
+
+
+def test_level_setting_resets_attributes():
+    ind = pd.MultiIndex.from_arrays([
+        ['A', 'A', 'B', 'B', 'B'], [1, 2, 1, 2, 3]
+    ])
+    assert ind.is_monotonic
+    ind.set_levels([['A', 'B'], [1, 3, 2]], inplace=True)
+    # if this fails, probably didn't reset the cache correctly.
+    assert not ind.is_monotonic
+
+
+def test_partial_string_timestamp_multiindex():
```

Comment:
partial string should be in test_indexing, or test_timeseries

Reply:
I moved this to test_indexing.